[rpc] add an option to record time spent waiting for GIL

rohan-varma opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#30842 [rpc] profile time spent blocked on GIL acquisition**

We'd like to profile the time spent on GIL acqusiition to debug
performance issues.

Differential Revision: [D18837590](https://our.internmc.facebook.com/intern/diff/D18837590/)

rohan-varma committed:
[wip][rpc] profile time spent blocked on GIL acquisition

We'd like to profile the time spent on GIL acqusiition to debug
performance issues.

Differential Revision: [D18837590](https://our.internmc.facebook.com/intern/diff/D18837590/)

[ghstack-poisoned]

mrshenli commented:
Is this only for testing or do plan to eventually land this?

satgera commented on torch/csrc/distributed/rpc/python_rpc_handler.cpp:
```diff
@@ -63,7 +64,12 @@ py::object PythonRpcHandler::loadPythonUDFResult(
   pybind11::gil_scoped_acquire ag;
+  auto dur = std::chrono::duration_cast<std::chrono::microseconds>(
+      std::chrono::high_resolution_clock::now() - start);
+  LOG(INFO) << "runPythonUDF: Took " << dur.count() << " us to acquire GIL.";
```
Oh I don't think we can add a LOG(INFO) here as its gonna be too too noisy. For now may be keep track of avg and then we can figure out how to expose this information.

rohan-varma commented:
@mrshenli It isn't for landing in this state, I just put this up temporarily to see what the wait times will look like. Will update this to a proper PR shortly.

rohan-varma committed:
Update on "[wip][rpc] profile time spent blocked on GIL acquisition"


We'd like to profile the time spent on GIL acqusiition to debug
performance issues.

Differential Revision: [D18837590](https://our.internmc.facebook.com/intern/diff/D18837590/)

[ghstack-poisoned]

rohan-varma committed:
Update on "[rpc] profile time spent blocked on GIL acquisition"


We'd like to profile the time spent on GIL acqusiition to debug
performance issues.

Differential Revision: [D18837590](https://our.internmc.facebook.com/intern/diff/D18837590/)

[ghstack-poisoned]

rohan-varma committed:
Update on "[rpc] profile time spent blocked on GIL acquisition"


We'd like to profile the time spent on GIL acqusiition to debug
performance issues.

Differential Revision: [D18837590](https://our.internmc.facebook.com/intern/diff/D18837590/)

[ghstack-poisoned]

rohan-varma committed:
Update on "[rpc] profile time spent blocked on GIL acquisition"


We'd like to profile the time spent on GIL acqusiition to debug
performance issues.

Differential Revision: [D18837590](https://our.internmc.facebook.com/intern/diff/D18837590/)

[ghstack-poisoned]

jjlilley commented:
The approach here is good.

The only question I'd ask (beyond my comment in ThriftRpcAgent in the fb review) is whether we want the future ability to completely disable this overhead. 

But I suspect we can decide this later... (since the cost of a couple of c++ timer calls is probably fairly low vs. the relatively high overhead of using Python for anything)

rohan-varma committed:
Update on "[rpc] profile time spent blocked on GIL acquisition"


We'd like to profile the time spent on GIL acqusiition to debug
performance issues.

Differential Revision: [D18837590](https://our.internmc.facebook.com/intern/diff/D18837590/)

[ghstack-poisoned]

rohan-varma committed:
Update on "[rpc] profile time spent blocked on GIL acquisition"


We'd like to profile the time spent on GIL acqusiition to debug
performance issues.

Differential Revision: [D18837590](https://our.internmc.facebook.com/intern/diff/D18837590/)

[ghstack-poisoned]

pritamdamania87 commented on torch/csrc/distributed/rpc/rpc_agent.cpp:
```diff
@@ -34,6 +34,18 @@ void RpcAgent::setDefaultRpcAgent(std::shared_ptr<RpcAgent> defaultRpcAgent) {
+
+void RpcAgent::addGilWaitTime(
+    const std::chrono::microseconds gilWaitTime /* unused */) {
+  // No-op default implementation
```
We should implement this in this PR, for now we can keep track of the avg time and add it to getMetrics? We should also have some unit tests associated with this PR.

rohan-varma replied:
I've implemented this with a histogram on the ThriftRpcAgent side, though for the process group agent, I guess we can implement it with just the average.

pritamdamania87 commented on torch/csrc/distributed/rpc/rpc_agent.h:
```diff
@@ -143,7 +143,10 @@ class TORCH_API RpcAgent {
+  virtual std::unordered_map<std::string, std::string> getDebugInfo();
+
+  // Add GIL wait time data point to metrics
+  virtual void addGilWaitTime(const std::chrono::microseconds gilWaitTime);
```
Can we make this private and make PythonRPCHandler a friend class?

rohan-varma replied:
This means that `GilWaitTimeGuard` where this method is called will have to be the friend class, or we can make `PythonRPCHandler` the friend class and make GilWaitTimeGuard a subclass of `PythonRpcHandler`. Is there a preferred approach?

rohan-varma replied:
I went with making `GilWaitTimeGuard` a class within `PythonRpcHandler`. If we realize that we need to use it outside of `PythonRpcHandler`, we can consider a different approach.

rohan-varma committed:
Update on "[rpc] profile time spent blocked on GIL acquisition"


We'd like to profile the time spent on GIL acqusiition to debug
performance issues.

Differential Revision: [D18837590](https://our.internmc.facebook.com/intern/diff/D18837590/)

[ghstack-poisoned]

rohan-varma committed:
Update on "[rpc] profile time spent blocked on GIL acquisition"


We'd like to profile the time spent on GIL acqusiition to debug
performance issues.

Differential Revision: [D18837590](https://our.internmc.facebook.com/intern/diff/D18837590/)

[ghstack-poisoned]

kostmo commented:
## CircleCI build failures summary
As of commit 0e476a1:
* **5/5** failures introduced in this PR



## Detailed failure analysis

One may explore the probable reasons each build failed interactively [on the Dr. CI website](https://dr.pytorch.org/commit-details.html?sha1=0e476a11ce95b151295959b8e766e16d0a252271).

### 2 new failures recognized by patterns
The following build failures don't appear to be due to upstream breakage:
#### [![See CircleCI build](https://avatars0.githubusercontent.com/ml/7?s=12)](https://circleci.com/gh/pytorch/pytorch/3917092) caffe2_onnx_py3_6_clang7_ubuntu16_04_build (1/2)
**Step:** "Build" ([details](https://dr.pytorch.org/build-details.html?build_id=72842673))
```
Dec 12 04:13:21 In file included from /var/lib/jenkins/workspace/torch/csrc/distributed/rpc/process_group_agent.h:3:
Dec 12 04:13:21 In file included from /var/lib/jenkins/workspace/c10/core/thread_pool.h:3:
Dec 12 04:13:21 In file included from /usr/bin/../lib/gcc/x86_64-linux-gnu/5.4.0/../../../../include/c++/5.4.0/condition_variable:39:
Dec 12 04:13:21 In file included from /usr/bin/../lib/gcc/x86_64-linux-gnu/5.4.0/../../../../include/c++/5.4.0/mutex:38:
Dec 12 04:13:21 In file included from /usr/bin/../lib/gcc/x86_64-linux-gnu/5.4.0/../../../../include/c++/5.4.0/tuple:39:
Dec 12 04:13:21 In file included from /usr/bin/../lib/gcc/x86_64-linux-gnu/5.4.0/../../../../include/c++/5.4.0/array:38:
Dec 12 04:13:21 In file included from /usr/bin/../lib/gcc/x86_64-linux-gnu/5.4.0/../../../../include/c++/5.4.0/stdexcept:39:
Dec 12 04:13:21 In file included from /usr/bin/../lib/gcc/x86_64-linux-gnu/5.4.0/../../../../include/c++/5.4.0/string:41:
Dec 12 04:13:21 In file included from /usr/bin/../lib/gcc/x86_64-linux-gnu/5.4.0/../../../../include/c++/5.4.0/bits/allocator.h:46:
Dec 12 04:13:21 In file included from /usr/bin/../lib/gcc/x86_64-linux-gnu/5.4.0/../../../../include/x86_64-linux-gnu/c++/5.4.0/bits/c++allocator.h:33:
Dec 12 04:13:21 /usr/bin/../lib/gcc/x86_64-linux-gnu/5.4.0/../../../../include/c++/5.4.0/ext/new_allocator.h:120:23: error: call to implicitly-deleted copy constructor of 'std::pair<const std::__cxx11::basic_string<char>, std::unique_ptr<torch::distributed::rpc::ProcessGroupAgent::AverageMetricsTracker, std::default_delete<torch::distributed::rpc::ProcessGroupAgent::AverageMetricsTracker> > >'
Dec 12 04:13:21         { ::new((void *)__p) _Up(std::forward<_Args>(__args)...); }
Dec 12 04:13:21                              ^   ~~~~~~~~~~~~~~~~~~~~~~~~~~~
Dec 12 04:13:21 /usr/bin/../lib/gcc/x86_64-linux-gnu/5.4.0/../../../../include/c++/5.4.0/bits/alloc_traits.h:530:8: note: in instantiation of function template specialization '__gnu_cxx::new_allocator<std::pair<const std::__cxx11::basic_string<char>, std::unique_ptr<torch::distributed::rpc::ProcessGroupAgent::AverageMetricsTracker, std::default_delete<torch::distributed::rpc::ProcessGroupAgent::AverageMetricsTracker> > > >::construct<std::pair<const std::__cxx11::basic_string<char>, std::unique_ptr<torch::distributed::rpc::ProcessGroupAgent::AverageMetricsTracker, std::default_delete<torch::distributed::rpc::ProcessGroupAgent::AverageMetricsTracker> > >, const std::pair<const std::__cxx11::basic_string<char>, std::unique_ptr<torch::distributed::rpc::ProcessGroupAgent::AverageMetricsTracker, std::default_delete<torch::distributed::rpc::ProcessGroupAgent::AverageMetricsTracker> > > &>' requested here
Dec 12 04:13:21         { __a.construct(__p, std::forward<_Args>(__args)...); }
Dec 12 04:13:21               ^
Dec 12 04:13:21 /usr/bin/../lib/gcc/x86_64-linux-gnu/5.4.0/../../../../include/c++/5.4.0/bits/hashtable_policy.h:1955:28: note: in instantiation of function template specialization 'std::allocator_traits<std::allocator<std::pair<const std::__cxx11::basic_string<char>, std::unique_ptr<torch::distributed::rpc::ProcessGroupAgent::AverageMetricsTracker, std::default_delete<torch::distributed::rpc::ProcessGroupAgent::AverageMetricsTracker> > > > >::construct<std::pair<const std::__cxx11::basic_string<char>, std::unique_ptr<torch::distributed::rpc::ProcessGroupAgent::AverageMetricsTracker, std::default_delete<torch::distributed::rpc::ProcessGroupAgent::AverageMetricsTracker> > >, const std::pair<const std::__cxx11::basic_string<char>, std::unique_ptr<torch::distributed::rpc::ProcessGroupAgent::AverageMetricsTracker, std::default_delete<torch::distributed::rpc::ProcessGroupAgent::AverageMetricsTracker> > > &>' requested here
Dec 12 04:13:21             __value_alloc_traits::construct(__a, __n->_M_valptr(),
Dec 12 04:13:21                                   ^
Dec 12 04:13:21 /usr/bin/../lib/gcc/x86_64-linux-gnu/5.4.0/../../../../include/c++/5.4.0/bits/hashtable_policy.h:180:16: note: in instantiation of function template specialization 'std::__detail::_Hashtable_alloc<std::allocator<std::__detail::_Hash_node<std::pair<const std::__cxx11::basic_string<char>, std::unique_ptr<torch::distributed::rpc::ProcessGroupAgent::AverageMetricsTracker, std::default_delete<torch::distributed::rpc::ProcessGroupAgent::AverageMetricsTracker> > >, true> > >::_M_allocate_node<const std::pair<const std::__cxx11::basic_string<char>, std::unique_ptr<torch::distributed::rpc::ProcessGroupAgent::AverageMetricsTracker, std::default_delete<torch::distributed::rpc::ProcessGroupAgent::AverageMetricsTracker> > > &>' requested here
Dec 12 04:13:21         { return _M_h._M_allocate_node(std::forward<_Arg>(__arg)); }
```
#### [![See CircleCI build](https://avatars0.githubusercontent.com/ml/7?s=12)](https://circleci.com/gh/pytorch/pytorch/3917078) pytorch_linux_xenial_py3_clang5_asan_build (2/2)
**Step:** "Build" ([details](https://dr.pytorch.org/build-details.html?build_id=72842906))
```
Dec 12 04:13:57 In file included from /var/lib/jenkins/workspace/torch/csrc/distributed/rpc/process_group_agent.h:3:
Dec 12 04:13:57 In file included from /var/lib/jenkins/workspace/c10/core/thread_pool.h:3:
Dec 12 04:13:57 In file included from /usr/bin/../lib/gcc/x86_64-linux-gnu/5.4.0/../../../../include/c++/5.4.0/condition_variable:39:
Dec 12 04:13:57 In file included from /usr/bin/../lib/gcc/x86_64-linux-gnu/5.4.0/../../../../include/c++/5.4.0/mutex:38:
Dec 12 04:13:57 In file included from /usr/bin/../lib/gcc/x86_64-linux-gnu/5.4.0/../../../../include/c++/5.4.0/tuple:39:
Dec 12 04:13:57 In file included from /usr/bin/../lib/gcc/x86_64-linux-gnu/5.4.0/../../../../include/c++/5.4.0/array:38:
Dec 12 04:13:57 In file included from /usr/bin/../lib/gcc/x86_64-linux-gnu/5.4.0/../../../../include/c++/5.4.0/stdexcept:39:
Dec 12 04:13:57 In file included from /usr/bin/../lib/gcc/x86_64-linux-gnu/5.4.0/../../../../include/c++/5.4.0/string:41:
Dec 12 04:13:57 In file included from /usr/bin/../lib/gcc/x86_64-linux-gnu/5.4.0/../../../../include/c++/5.4.0/bits/allocator.h:46:
Dec 12 04:13:57 In file included from /usr/bin/../lib/gcc/x86_64-linux-gnu/5.4.0/../../../../include/x86_64-linux-gnu/c++/5.4.0/bits/c++allocator.h:33:
Dec 12 04:13:57 /usr/bin/../lib/gcc/x86_64-linux-gnu/5.4.0/../../../../include/c++/5.4.0/ext/new_allocator.h:120:23: error: call to implicitly-deleted copy constructor of 'std::pair<const std::__cxx11::basic_string<char>, std::unique_ptr<torch::distributed::rpc::ProcessGroupAgent::AverageMetricsTracker, std::default_delete<torch::distributed::rpc::ProcessGroupAgent::AverageMetricsTracker> > >'
Dec 12 04:13:57         { ::new((void *)__p) _Up(std::forward<_Args>(__args)...); }
Dec 12 04:13:57                              ^   ~~~~~~~~~~~~~~~~~~~~~~~~~~~
Dec 12 04:13:57 /usr/bin/../lib/gcc/x86_64-linux-gnu/5.4.0/../../../../include/c++/5.4.0/bits/alloc_traits.h:530:8: note: in instantiation of function template specialization '__gnu_cxx::new_allocator<std::pair<const std::__cxx11::basic_string<char>, std::unique_ptr<torch::distributed::rpc::ProcessGroupAgent::AverageMetricsTracker, std::default_delete<torch::distributed::rpc::ProcessGroupAgent::AverageMetricsTracker> > > >::construct<std::pair<const std::__cxx11::basic_string<char>, std::unique_ptr<torch::distributed::rpc::ProcessGroupAgent::AverageMetricsTracker, std::default_delete<torch::distributed::rpc::ProcessGroupAgent::AverageMetricsTracker> > >, const std::pair<const std::__cxx11::basic_string<char>, std::unique_ptr<torch::distributed::rpc::ProcessGroupAgent::AverageMetricsTracker, std::default_delete<torch::distributed::rpc::ProcessGroupAgent::AverageMetricsTracker> > > &>' requested here
Dec 12 04:13:57         { __a.construct(__p, std::forward<_Args>(__args)...); }
Dec 12 04:13:57               ^
Dec 12 04:13:57 /usr/bin/../lib/gcc/x86_64-linux-gnu/5.4.0/../../../../include/c++/5.4.0/bits/hashtable_policy.h:1955:28: note: in instantiation of function template specialization 'std::allocator_traits<std::allocator<std::pair<const std::__cxx11::basic_string<char>, std::unique_ptr<torch::distributed::rpc::ProcessGroupAgent::AverageMetricsTracker, std::default_delete<torch::distributed::rpc::ProcessGroupAgent::AverageMetricsTracker> > > > >::construct<std::pair<const std::__cxx11::basic_string<char>, std::unique_ptr<torch::distributed::rpc::ProcessGroupAgent::AverageMetricsTracker, std::default_delete<torch::distributed::rpc::ProcessGroupAgent::AverageMetricsTracker> > >, const std::pair<const std::__cxx11::basic_string<char>, std::unique_ptr<torch::distributed::rpc::ProcessGroupAgent::AverageMetricsTracker, std::default_delete<torch::distributed::rpc::ProcessGroupAgent::AverageMetricsTracker> > > &>' requested here
Dec 12 04:13:57             __value_alloc_traits::construct(__a, __n->_M_valptr(),
Dec 12 04:13:57                                   ^
Dec 12 04:13:57 /usr/bin/../lib/gcc/x86_64-linux-gnu/5.4.0/../../../../include/c++/5.4.0/bits/hashtable_policy.h:180:16: note: in instantiation of function template specialization 'std::__detail::_Hashtable_alloc<std::allocator<std::__detail::_Hash_node<std::pair<const std::__cxx11::basic_string<char>, std::unique_ptr<torch::distributed::rpc::ProcessGroupAgent::AverageMetricsTracker, std::default_delete<torch::distributed::rpc::ProcessGroupAgent::AverageMetricsTracker> > >, true> > >::_M_allocate_node<const std::pair<const std::__cxx11::basic_string<char>, std::unique_ptr<torch::distributed::rpc::ProcessGroupAgent::AverageMetricsTracker, std::default_delete<torch::distributed::rpc::ProcessGroupAgent::AverageMetricsTracker> > > &>' requested here
Dec 12 04:13:57         { return _M_h._M_allocate_node(std::forward<_Arg>(__arg)); }
```
### 3 failures *not* recognized by patterns:
| Job | Step |
| --- | --- |
| ![CircleCI](https://avatars0.githubusercontent.com/ml/7?s=12) <sup>pytorch_linux_xenial_cuda9_cudnn7_py3_build</sup> | <sup>Build</sup> |
| ![CircleCI](https://avatars0.githubusercontent.com/ml/7?s=12) <sup>pytorch_linux_xenial_py3_6_gcc5_4_build</sup> | <sup>Build</sup> |
| ![CircleCI](https://avatars0.githubusercontent.com/ml/7?s=12) <sup>pytorch_xla_linux_xenial_py3_6_clang7_build</sup> | <sup>Build</sup> |


---
This comment was automatically generated by [Dr. CI](https://github.com/kostmo/circleci-failure-tracker/tree/master/docs/from-pull-request-comment).
Follow [this link to opt-out](https://dr.pytorch.org/admin/comments-opt-out.html) of these comments for your Pull Requests.


Please report bugs/suggestions on the [GitHub issue tracker](https://github.com/kostmo/circleci-failure-tracker/issues).

rohan-varma committed:
Update on "[rpc] profile time spent blocked on GIL acquisition"


We'd like to profile the time spent on GIL acqusiition to debug
performance issues.

Differential Revision: [D18837590](https://our.internmc.facebook.com/intern/diff/D18837590/)

[ghstack-poisoned]

jjlilley commented on torch/csrc/distributed/rpc/process_group_agent.h:
```diff
@@ -88,6 +87,30 @@ class ProcessGroupAgent : public RpcAgent {
+  // to track num_sends, recvs, average size of messages, etc.
+  struct AverageMetricsTracker {
+    const std::string key_;
+    float currentAverage_;
```
misc: unless you have zillions of these, maybe better to use double/int64 rather than float/int32?

float32 precision isn't so high, can decay quickly in practice, especially if you have values that are far apart (e.g. below (x + (y-x)/ 1 million) might start not modifying the value much)

also: maybe consider making the representation {sum, count} rather than {average, count}?

rohan-varma replied:
We may be able to do this, though if we change average to sum, there could be overflow issues more easily, depending on how long-running a processes that writes to these metrics are.

jjlilley commented on torch/csrc/distributed/rpc/process_group_agent.cpp:
```diff
@@ -568,17 +572,28 @@ std::unordered_map<std::string, std::string> ProcessGroupAgent::getMetrics() {
-  return getMetrics();
+void ProcessGroupAgent::addGilWaitTime(
+    const std::chrono::microseconds gilWaitTime) {
+  auto gilMetrics = metricsMap_.find(kGilAverageWaitTime);
```
need to protect the map with a lock. This is a bit of work to do on each GIL acquisition... also, not sure how it dovetails with other work people have done in this area.

Since this is the only use case of the map, my preference here would be just to keep a standalone std::unique_ptr<> just for this, and avoid an expensive string-based table lookup for it each time. 

If you eventually want it to be more general/etc. you can just keep vector<unique_ptr<AverageMetricsTracker>>, where you have an integer-like stats enum (e.g. AVERAGE_WAIT_TIME=0, SOME_OTHER_STAT=1,...) to index into it, rather than string-based indexing. (It still is "general" since you have the key name in the struct, and can emit it in the metrics method.)

rohan-varma replied:
Implemented this using a `vector<unique_ptr>` for now.

rohan-varma committed:
Update on "[rpc] profile time spent blocked on GIL acquisition"


We'd like to profile the time spent on GIL acqusiition to debug
performance issues.

Differential Revision: [D18837590](https://our.internmc.facebook.com/intern/diff/D18837590/)

[ghstack-poisoned]

rohan-varma committed:
Update on "[rpc] profile time spent blocked on GIL acquisition"


We'd like to profile the time spent on GIL acqusiition to debug
performance issues.

Differential Revision: [D18837590](https://our.internmc.facebook.com/intern/diff/D18837590/)

[ghstack-poisoned]

rohan-varma commented on torch/csrc/distributed/rpc/process_group_agent.h:
```diff
@@ -88,6 +87,30 @@ class ProcessGroupAgent : public RpcAgent {
     std::mutex mutex_;
   };
 
+  // TODO: this class should inherit from a MetricsTracker, and can be extended
```
@jjlilley, if I do this and add the forward declaration here, then I'm getting an error "member access into an incomplete type" when I try to use it in `addGilWaitTime` in process group agent. Is there any workaround for this? For now, I changed it to define the struct here but the implementations in the cpp file.

rohan-varma committed:
Update on "[rpc] profile time spent blocked on GIL acquisition"


We'd like to profile the time spent on GIL acqusiition to debug
performance issues.

Differential Revision: [D18837590](https://our.internmc.facebook.com/intern/diff/D18837590/)

[ghstack-poisoned]

rohan-varma committed:
Update on "[rpc] profile time spent blocked on GIL acquisition"


We'd like to profile the time spent on GIL acqusiition to debug
performance issues.

Differential Revision: [D18837590](https://our.internmc.facebook.com/intern/diff/D18837590/)

[ghstack-poisoned]

rohan-varma committed:
Update on "[rpc] profile time spent blocked on GIL acquisition"


We'd like to profile the time spent on GIL acqusiition to debug
performance issues.

Differential Revision: [D18837590](https://our.internmc.facebook.com/intern/diff/D18837590/)

[ghstack-poisoned]

rohan-varma committed:
Update on "[rpc] profile time spent blocked on GIL acquisition"


We'd like to profile the time spent on GIL acqusiition to debug
performance issues.

Differential Revision: [D18837590](https://our.internmc.facebook.com/intern/diff/D18837590/)

[ghstack-poisoned]

satgera reviewed (approved):
Looks good !

jjlilley reviewed (commented):
Thanks for working on this!
I have no objections as-is, though I added a few comments on the Phab diff D18837590

rohan-varma committed:
Update on "[rpc] profile time spent blocked on GIL acquisition"


We'd like to profile the time spent on GIL acqusiition to debug
performance issues.

Differential Revision: [D18837590](https://our.internmc.facebook.com/intern/diff/D18837590/)

[ghstack-poisoned]

jjlilley commented:
Thanks for updating per my Phabricator comments! [sorry, I thought I'd left them here], lgtm.

pritamdamania87 commented on torch/csrc/distributed/rpc/python_rpc_handler.h:
```diff
@@ -61,6 +61,15 @@ class PYBIND11_EXPORT PythonRpcHandler {
+    void markAcquired();
+
+   private:
+    std::chrono::time_point<std::chrono::high_resolution_clock> start;
```
nit: `start_`.

pritamdamania87 commented on torch/csrc/distributed/rpc/python_rpc_handler.cpp:
```diff
@@ -56,33 +72,43 @@ std::vector<char> PythonRpcHandler::generatePythonUDFResult(
     const SerializedPyObj& serializedObj) {
+  GilWaitTimeGuard g;
   pybind11::gil_scoped_acquire ag;
+  g.markAcquired();
```
Can we make this optional such that it can be enabled/disabled via flag (default can be enabled for now)? There is quite a bit of overhead in `markAcquired`. Also, this block of code could be replaced with a MACRO (which performs appropriate guards based on whether or not this feature is enabled) instead of repeating these three lines everywhere.

rohan-varma replied:
@pritamdamania87 Did you mean a compile-time flag or a runtime variable that can be toggled?

rohan-varma replied:
@pritamdamania87 Went ahead and made it configurable at runtime.

rohan-varma committed:
Update on "[rpc] profile time spent blocked on GIL acquisition"


We'd like to profile the time spent on GIL acqusiition to debug
performance issues.

Differential Revision: [D18837590](https://our.internmc.facebook.com/intern/diff/D18837590/)

[ghstack-poisoned]

rohan-varma committed:
Update on "[rpc] profile time spent blocked on GIL acquisition"


We'd like to profile the time spent on GIL acqusiition to debug
performance issues.

Differential Revision: [D18837590](https://our.internmc.facebook.com/intern/diff/D18837590/)

[ghstack-poisoned]

rohan-varma committed:
Update on "[rpc] profile time spent blocked on GIL acquisition"


We'd like to profile the time spent on GIL acqusiition to debug
performance issues.

Differential Revision: [D18837590](https://our.internmc.facebook.com/intern/diff/D18837590/)

[ghstack-poisoned]

rohan-varma committed:
Update on "[rpc] profile time spent blocked on GIL acquisition"


We'd like to profile the time spent on GIL acqusiition to debug
performance issues.

Differential Revision: [D18837590](https://our.internmc.facebook.com/intern/diff/D18837590/)

[ghstack-poisoned]

rohan-varma committed:
Update on "[rpc] profile time spent blocked on GIL acquisition"


We'd like to profile the time spent on GIL acqusiition to debug
performance issues.

Differential Revision: [D18837590](https://our.internmc.facebook.com/intern/diff/D18837590/)

[ghstack-poisoned]

rohan-varma committed:
Update on "[rpc] profile time spent blocked on GIL acquisition"


We'd like to profile the time spent on GIL acqusiition to debug
performance issues.

Differential Revision: [D18837590](https://our.internmc.facebook.com/intern/diff/D18837590/)

[ghstack-poisoned]

rohan-varma committed:
Update on "[rpc] profile time spent blocked on GIL acquisition"


We'd like to profile the time spent on GIL acqusiition to debug
performance issues.

Differential Revision: [D18837590](https://our.internmc.facebook.com/intern/diff/D18837590/)

[ghstack-poisoned]

pritamdamania87 commented on torch/csrc/distributed/rpc/python_rpc_handler.h:
```diff
@@ -61,6 +61,24 @@ class PYBIND11_EXPORT PythonRpcHandler {
+  pybind11::gil_scoped_acquire ag;                             \
+  if (RpcAgent::getDefaultRpcAgent()->getMetricsProfiling()) { \
+    g.markAcquired();                                          \
+  }
```
Just realized this is not really a guard since we don't actually do anything when the object is destroyed. Can we change the macro to do this instead:

```
std::chrono::time_point<std::chrono::high_resolution_clock> startTime;
bool should_profile_gil = RpcAgent::getDefaultRpcAgent()->getMetricsProfiling();
if (should_profile_gil) {
  startTime = now();
}
pybind11::gil_scoped_acquire ag;
if (should_profile_gil) {
  auto gil_time = std::chrono::duration_cast<std::chrono::microseconds>(
      std::chrono::high_resolution_clock::now() - start);
  RpcAgent::getDefaultRpcAgent()->addGilWaitTime(gil_time);
}
```

This would minimize the overhead when GIL profiling is disabled and we probably don't need a separate GilWaitTimeGuard class since the macro handles this.

rohan-varma commented on torch/csrc/distributed/rpc/process_group_agent.cpp:
```diff
@@ -50,10 +50,10 @@ double ProcessGroupAgent::AverageMetricsTracker::computeAverage() {
-const std::string kThreadPoolSize = "thread_pool_size";
-const std::string kNumIdleThreads = "num_idle_threads";
-const std::string kGilAverageWaitTime = "gil_average_wait_time_us";
+const std::string kNumPendingRequests = "agent.num_pending_requests";
```
@pritamdamania87 I'm trying to keep it the same as the metrics for the internal RPC Agent which just uses "agent" as well. Would it be better to rename both of them?

pritamdamania87 replied:
I see, I guess "agent" makes sense since we would actually only use 1 agent at a time. Although, maybe we can rename "agent" to "rpc_agent" in a future diff for more clarity.

rohan-varma commented on test/rpc_test.py:
```diff
@@ -1247,6 +1247,21 @@ def test_rref_context_debug_info(self):
+        # test that rpc.set_metrics_profiling(false) will result in more
+        # expensive metrics such as GIL wait time not being recorded.
+        from torch.distributed.rpc.api import _agent
+        rpc.set_metrics_profiling(False)
```
I was planning to make it more general once we start working on the metrics stuff, although I don't think it makes sense to have this as `metrics` if there's only GIL profiling for now. Let's just make it GIL specific for now and we can change later as needed.

rohan-varma committed:
Update on "[rpc] profile time spent blocked on GIL acquisition"


We'd like to profile the time spent on GIL acqusiition to debug
performance issues.

Differential Revision: [D18837590](https://our.internmc.facebook.com/intern/diff/D18837590/)

[ghstack-poisoned]

rohan-varma committed:
Update on "[rpc] profile time spent blocked on GIL acquisition"


We'd like to profile the time spent on GIL acqusiition to debug
performance issues.

Differential Revision: [D18837590](https://our.internmc.facebook.com/intern/diff/D18837590/)

[ghstack-poisoned]

rohan-varma commented on test/rpc_test.py:
```diff
@@ -1303,20 +1303,37 @@ def test_rref_context_debug_info(self):
+
     @dist_init
     @requires_process_group_agent("PROCESS_GROUP rpc backend specific test, skip")
     def test_process_group_debug_info(self):
```
Filed https://github.com/pytorch/pytorch/issues/32362 to ensure that we eventually make this test backend agnostic.

rohan-varma committed:
Update on "[rpc] profile time spent blocked on GIL acquisition"


We'd like to profile the time spent on GIL acqusiition to debug
performance issues.

Differential Revision: [D18837590](https://our.internmc.facebook.com/intern/diff/D18837590/)

[ghstack-poisoned]

rohan-varma committed:
Update on "[rpc] profile time spent blocked on GIL acquisition"


We'd like to profile the time spent on GIL acqusiition to debug
performance issues.

Differential Revision: [D18837590](https://our.internmc.facebook.com/intern/diff/D18837590/)

[ghstack-poisoned]

rohan-varma committed:
Update on "[rpc] profile time spent blocked on GIL acquisition"


We'd like to profile the time spent on GIL acqusiition to debug
performance issues.

Differential Revision: [D18837590](https://our.internmc.facebook.com/intern/diff/D18837590/)

[ghstack-poisoned]

rohan-varma committed:
Update on "[rpc] profile time spent blocked on GIL acquisition"


We'd like to profile the time spent on GIL acqusiition to debug
performance issues.

Differential Revision: [D18837590](https://our.internmc.facebook.com/intern/diff/D18837590/)

[ghstack-poisoned]
